include/seqan3/core/bit_manipulation.hpp

```diff
@@ -77,30 +79,159 @@ constexpr size_t next_power_of_two(size_t n)
+*
+* ### Complexity
+*
+* Constant.
+*/
+template <std::UnsignedIntegral unsigned_t>
+constexpr uint8_t popcount(unsigned_t const n) noexcept
+{
+#if defined(__GNUC__)
+    if constexpr (sizeof(unsigned_t) == sizeof(unsigned long long))
+        return __builtin_popcountll(n);
+    else if constexpr (sizeof(unsigned_t) == sizeof(unsigned long))
+        return __builtin_popcountl(n);
+    else
+        return __builtin_popcount(n);
+#else
+    return sdsl::bits::cnt(n);
+#endif
+}
+
+/*!\brief Returns the number of leading 0-bits, starting at the most significant bit position.
+* \ingroup core
+*
+* \param[in] n An unsigned integer.
+*
+* \attention *n = 0* is a special case and is undefined behaviour.
+*
+* \returns The number of leading 0-bits.
+*
+* ### Example
+*
+* \include test/snippet/core/detail/count_leading_zeros.cpp
+*
+* ### Exception
+*
+* No-throw guarantee.
+*
+* ### Thread-safety
+*
+* Thread safe.
+*
+* ### Complexity
+*
+* Constant.
+*/
+template <std::UnsignedIntegral unsigned_t>
+constexpr uint8_t count_leading_zeros(unsigned_t const n) noexcept
+{
+    assert(n > 0); // n == 0 might have undefined behaviour
+#if defined(__GNUC__)
+    if constexpr (sizeof(unsigned_t) == sizeof(unsigned long long))
+        return __builtin_clzll(n);
+    else if constexpr (sizeof(unsigned_t) == sizeof(unsigned long))
+        return __builtin_clzl(n);
+    else
+        return __builtin_clz(n) + sizeof_bits<unsigned_t> - sizeof_bits<unsigned int>;
+#else
+    return sizeof_bits<unsigned_t> - sdsl::bits::hi(n) - 1;
+#endif
+}
+
+/*!\brief Returns the number of trailing 0-bits, starting at the least significant bit position.
  * \ingroup core
  *
  * \param[in] n An unsigned integer.
  *
  * \attention *n = 0* is a special case and is undefined behaviour.
  *
- * \returns The position of the most significant bit.
+ * \returns The number of trailing 0-bits.
+ *
+ * ### Example
+ *
+ * \include test/snippet/core/detail/count_trailing_zeros.cpp
+ *
+ * ### Exception
+ *
+ * No-throw guarantee.
+ *
+ * ### Thread-safety
+ *
+ * Thread safe.
+ *
+ * ### Complexity
+ *
+ * Constant.
  */
 template <std::UnsignedIntegral unsigned_t>
-constexpr uint8_t bit_scan_reverse(unsigned_t n)
+constexpr uint8_t count_trailing_zeros(unsigned_t const n) noexcept
 {
     assert(n > 0); // n == 0 might have undefined behaviour
 #if defined(__GNUC__)
     if constexpr (sizeof(unsigned_t) == sizeof(unsigned long long))
-        return sizeof_bits<unsigned long long> - __builtin_clzll(n) - 1;
+        return __builtin_ctzll(n);
     else if constexpr (sizeof(unsigned_t) == sizeof(unsigned long))
-        return sizeof_bits<unsigned long> - __builtin_clzl(n) - 1;
+        return __builtin_ctzl(n);
     else
-        return sizeof_bits<unsigned> - __builtin_clz(n) - 1;
+        return __builtin_ctz(n);
+#else
+    return sdsl::bits::lo(n);
+#endif
+}
+
+/*!\brief Returns the position (0-based) of the most significant bit, starting at the least significant bit position.
```

Comment:
but it scans from msb to lsb, doesn't it?

Reply:
Mhm, it's not really "scanning", or what do you mean?
`starting at the least significant bit position` refers to the positions, i.e. we count from right to left.

Reply:
Ok, you want to indicate that the position is the postion when reading from right to left. But I think it is confusing. I think the first part of the sentence is clear enough about what will be returned. Also there is an example so it is clear how it is used. Maybe just drop the second part then.